File: backend/app/models/vitality_report.py
"""Vitality report storage model."""

import uuid
from datetime import datetime

from sqlalchemy import DateTime, Float, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from app.database import Base


class VitalityReport(Base):
    __tablename__ = "vitality_reports"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    user_id: Mapped[str] = mapped_column(String(36), nullable=False, index=True)

    # Report window
    period_start: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    period_end: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    scan_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    alert_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)

    # Metric averages (current week)
    avg_hr_bpm: Mapped[float | None] = mapped_column(Float, nullable=True)
    avg_hrv_ms: Mapped[float | None] = mapped_column(Float, nullable=True)
    avg_respiratory_rate: Mapped[float | None] = mapped_column(Float, nullable=True)
    avg_voice_jitter_pct: Mapped[float | None] = mapped_column(Float, nullable=True)
    avg_voice_shimmer_pct: Mapped[float | None] = mapped_column(Float, nullable=True)

    # Week-over-week deltas
    delta_hr_bpm: Mapped[float | None] = mapped_column(Float, nullable=True)
    delta_hrv_ms: Mapped[float | None] = mapped_column(Float, nullable=True)

    # Supplementary indicators (latest from the period)
    latest_vascular_age_estimate: Mapped[float | None] = mapped_column(Float, nullable=True)
    latest_vascular_age_confidence: Mapped[float | None] = mapped_column(Float, nullable=True)
    latest_anemia_label: Mapped[str | None] = mapped_column(String(32), nullable=True)
    latest_anemia_confidence: Mapped[float | None] = mapped_column(Float, nullable=True)

    # Rendered text summary
    summary_text: Mapped[str] = mapped_column(Text, nullable=False, default="")

    generated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )

File: service-core/src/main/java/com/pranapulse/core/report/domain/VitalityReport.java
package com.pranapulse.core.report.domain;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Index;
import jakarta.persistence.PrePersist;
import jakarta.persistence.Table;
import java.time.Instant;
import java.util.UUID;

@Entity
@Table(
        schema = "core",
        name = "vitality_reports",
        indexes = {
                @Index(name = "idx_vitality_reports_user_id", columnList = "user_id"),
                @Index(name = "idx_vitality_reports_generated_at", columnList = "generated_at")
        }
)
public class VitalityReport {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(name = "user_id", nullable = false)
    private UUID userId;

    @Column(name = "period_start", nullable = false)
    private Instant periodStart;

    @Column(name = "period_end", nullable = false)
    private Instant periodEnd;

    @Column(name = "scan_count", nullable = false)
    private int scanCount;

    @Column(name = "alert_count", nullable = false)
    private int alertCount;

    @Column(name = "avg_hr_bpm")
    private Double avgHrBpm;

    @Column(name = "avg_hrv_ms")
    private Double avgHrvMs;

    @Column(name = "avg_respiratory_rate")
    private Double avgRespiratoryRate;

    @Column(name = "avg_voice_jitter_pct")
    private Double avgVoiceJitterPct;

    @Column(name = "avg_voice_shimmer_pct")
    private Double avgVoiceShimmerPct;

    @Column(name = "delta_hr_bpm")
    private Double deltaHrBpm;

    @Column(name = "delta_hrv_ms")
    private Double deltaHrvMs;

    @Column(name = "latest_vascular_age_estimate")
    private Double latestVascularAgeEstimate;

    @Column(name = "latest_vascular_age_confidence")
    private Double latestVascularAgeConfidence;

    @Column(name = "latest_anemia_label", length = 32)
    private String latestAnemiaLabel;

    @Column(name = "latest_anemia_confidence")
    private Double latestAnemiaConfidence;

    @Column(name = "summary_text", nullable = false, length = 4000)
    private String summaryText;

    @Column(name = "generated_at", nullable = false, updatable = false)
    private Instant generatedAt;

    protected VitalityReport() {
    }

    public VitalityReport(
            UUID userId,
            Instant periodStart,
            Instant periodEnd,
            int scanCount,
            int alertCount,
            Double avgHrBpm,
            Double avgHrvMs,
            Double avgRespiratoryRate,
            Double avgVoiceJitterPct,
            Double avgVoiceShimmerPct,
            Double deltaHrBpm,
            Double deltaHrvMs,
            Double latestVascularAgeEstimate,
            Double latestVascularAgeConfidence,
            String latestAnemiaLabel,
            Double latestAnemiaConfidence,
            String summaryText
    ) {
        this.userId = userId;
        this.periodStart = periodStart;
        this.periodEnd = periodEnd;
        this.scanCount = scanCount;
        this.alertCount = alertCount;
        this.avgHrBpm = avgHrBpm;
        this.avgHrvMs = avgHrvMs;
        this.avgRespiratoryRate = avgRespiratoryRate;
        this.avgVoiceJitterPct = avgVoiceJitterPct;
        this.avgVoiceShimmerPct = avgVoiceShimmerPct;
        this.deltaHrBpm = deltaHrBpm;
        this.deltaHrvMs = deltaHrvMs;
        this.latestVascularAgeEstimate = latestVascularAgeEstimate;
        this.latestVascularAgeConfidence = latestVascularAgeConfidence;
        this.latestAnemiaLabel = latestAnemiaLabel;
        this.latestAnemiaConfidence = latestAnemiaConfidence;
        this.summaryText = summaryText;
    }

    @PrePersist
    void onCreate() {
        if (generatedAt == null) {
            generatedAt = Instant.now();
        }
    }

    public UUID getId() {
        return id;
    }

    public UUID getUserId() {
        return userId;
    }

    public Instant getPeriodStart() {
        return periodStart;
    }

    public Instant getPeriodEnd() {
        return periodEnd;
    }

    public int getScanCount() {
        return scanCount;
    }

    public int getAlertCount() {
        return alertCount;
    }

    public Double getAvgHrBpm() {
        return avgHrBpm;
    }

    public Double getAvgHrvMs() {
        return avgHrvMs;
    }

    public Double getAvgRespiratoryRate() {
        return avgRespiratoryRate;
    }

    public Double getAvgVoiceJitterPct() {
        return avgVoiceJitterPct;
    }

    public Double getAvgVoiceShimmerPct() {
        return avgVoiceShimmerPct;
    }

    public Double getDeltaHrBpm() {
        return deltaHrBpm;
    }

    public Double getDeltaHrvMs() {
        return deltaHrvMs;
    }

    public Double getLatestVascularAgeEstimate() {
        return latestVascularAgeEstimate;
    }

    public Double getLatestVascularAgeConfidence() {
        return latestVascularAgeConfidence;
    }

    public String getLatestAnemiaLabel() {
        return latestAnemiaLabel;
    }

    public Double getLatestAnemiaConfidence() {
        return latestAnemiaConfidence;
    }

    public String getSummaryText() {
        return summaryText;
    }

    public Instant getGeneratedAt() {
        return generatedAt;
    }
}
